keps/sig-apps/20190326-stateful-application-api.md

```diff
@@ -0,0 +1,790 @@
+	// taking an application snapshot. The hooks may return application-specific tags in JSON
+	// formatted key-value pairs, e.g., {“tag1”: “value1”, “tag2”: “value2”}. Each hook in the
+	// list is for a specific container of the application, and one container can have no more
+	// than one pre-hook.
+	// +optional
+	PreHooks []HookSpec `json:"preHooks,omitempty"`
+
+	// PostHooks are an optional list of hooks  to run to do application unquiescing after taking
+	// an application snapshot. Each hook in the list is for a specific container of the application,
+	// and one container of the application can have no more than one post-hook.
+	// +optional
+	PostHooks []HookSpec `json:"postHooks,omitempty"`
+
+	// DeletionPolicy is the policy on what happens to an ApplicationSnapshotContent resource and the
+	// local application snapshot it represents when the ApplicationSnapshot resource it is bound to
+	// is deleted.
+	// DeletionPolicy set in individual ApplicationSnapshotContent resources takes precedence over this.
+	// +optional
+	// +kubebuilder:validation:Enum=Delete,Retain
+	DeletionPolicy *DeletionPolicy `json:"deletionPolicy,omitempty"`
+}
+
+// HookSpec defines a hook for quiescing or unquiescing an application.
+type HookSpec struct {
+	// Name is the name of the hook.
+	Name string `json:"name"`
+
+	// PodFilter is an optional label selector for filtering the pods to execute the hook in.
+	// +optional
+	PodFilter *metav1.LabelSelector `json:"podFilter,omitempty"`
+
+	// Container specifies the name of the container to execute the hook in. If not specified, the first
+	// container is selected.
+	// +optional
+	Container *string `json:"container,omitempty"`
+
+  // ActionName specifies the name of the HookAction that defines the action to take for this hook. 
+  // The HookAction resource type is proposed in https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190120-execution-hook-design.md#implementation-history.
+	ActionName string `json:"actionName,omitempty"`
+}
+
+// DeletionPolicy defines the policy on what happens to an ApplicationSnapshotContent resource and the
+// local application snapshot it represents when the ApplicationSnapshot resource it is bound to is deleted.
+type DeletionPolicy string
+
+// Set of DeletionPolicy values.
+const (
+	DeletionPolicyDelete DeletionPolicy = "Delete"
+	DeletionPolicyRetain DeletionPolicy = "Retain"
+)
+```
+
+`ApplicationSnapshotContent` is a cluster-scoped CRD that describes the details about a local application snapshot. An `ApplicationSnapshotContent` object is created as the result of successfully processing an `ApplicationSnapshot`. The relationship between an `ApplicationSnapshot` object and an `ApplicationSnapshotContent` object is like the relationship between a `VolumeSnapshot` object and a `VolumeSnapshotContent` object. Below is the definition of `ApplicationSnapshotContent`.
+
```

Comment:
How is the binding of individual VolumeSnapshot/VolumeSnapshotContent related to the binding of ApplicationSnapshot/ApplicationSnapshotContent?  Is static provisioning supported for ApplicationSnapshot?

Reply:
We support static binding of an `ApplicationSnapshot` to an existing `ApplicationSnapshotContent`.